### PR TITLE
VMware: New Module: vmware_host_scanhba

### PR DESCRIPTION
**Pull request by mikeeaton83**
##### SUMMARY
vmware_host_scanhba

Scan the HBAs on an esxi host and optionally refresh the storage system as required.

##### ISSUE TYPE

- New Module Pull Request

##### COMPONENT NAME
vmware_host_scanhba

##### ADDITIONAL INFORMATION
When using the vmware_host_datastore module to mount a new datastore that has been presented to a esxi host you need to rescan the storage system for esxi to see this new volume, hence this module.

### CODE REVIEW

File: lib/ansible/modules/cloud/vmware/vmware_host_scanhba.py [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+}
+
+DOCUMENTATION = r'''
+---
+module: vmware_host_scanhba
+short_description: Rescan host HBA's and optionally refresh the storage system
+description:
+- This module can force a rescan of the hosts HBA subsystem which is needed when wanting to mount a new datastore.
+- You could use this before using vmware_host_datastore to mount a new datastore to ensure your device/volume is ready.
+- You can also optionally force a Refresh of the Storage System in vCenter/ESXi Web Client.
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/vmware/vmware_host_scanhba.py:24]:
```suggestion
- You can also optionally force refresh operation on the Storage System in vCenter/ESXi Web Client.
```

File: lib/ansible/modules/cloud/vmware/vmware_host_scanhba.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+    'supported_by': 'community'
+}
+
+DOCUMENTATION = r'''
+---
+module: vmware_host_scanhba
+short_description: Rescan host HBA's and optionally refresh the storage system
+description:
+- This module can force a rescan of the hosts HBA subsystem which is needed when wanting to mount a new datastore.
+- You could use this before using vmware_host_datastore to mount a new datastore to ensure your device/volume is ready.
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/vmware/vmware_host_scanhba.py:23]:
```suggestion
- You could use this before using M(vmware_host_datastore) to mount a new datastore to ensure your device/volume is ready.
```

File: lib/ansible/modules/cloud/vmware/vmware_host_scanhba.py [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+- python >= 2.6
+- PyVmomi
+options:
+  esxi_hostname:
+    description:
+    - ESXi hostname to Rescan the storage subsystem on.
+    required: true
+  refresh_storage:
+    description:
+    - Refresh the storage system in vCenter/ESXi Web Client
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/vmware/vmware_host_scanhba.py]:
```suggestion
    - Refresh the storage system in vCenter/ESXi Web Client.
```

File: lib/ansible/modules/cloud/vmware/vmware_host_scanhba.py [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+
+EXAMPLES = r'''
+- name: Recan HBA's for a given ESXi host and refresh storage system objects
+  vmware_host_scanhba:
+      hostname: '{{ vcenter_hostname }}'
+      username: '{{ vcenter_username }}'
+      password: '{{ vcenter_password }}'
+      esxi_hostname: '{{ inventory_hostname }}'
+      refresh_storage: true
+  delegate_to: localhost
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/vmware/vmware_host_scanhba.py:61]: Add a blank after this to notify that these are two separate tasks.

File: lib/ansible/modules/cloud/vmware/vmware_host_scanhba.py [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+- name: Recan HBA's for a given ESXi host and don't refresh storage system objects
+  vmware_host_scanhba:
+      hostname: '{{ vcenter_hostname }}'
+      username: '{{ vcenter_username }}'
+      password: '{{ vcenter_password }}'
+      esxi_hostname: '{{ inventory_hostname }}'
+  delegate_to: localhost
+'''
+
+RETURN = r'''
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/vmware/vmware_host_scanhba.py:81]: It is not mandatory to have return but I would recommend to add one, from UI/UX point. :)

File: lib/ansible/modules/cloud/vmware/vmware_host_scanhba.py [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+
+
+class VmwareHbaScan(PyVmomi):
+    def __init__(self, module):
+        super(VmwareHbaScan, self).__init__(module)
+
+    def scan(self):
+        esxi_host_name = self.params.get('esxi_hostname', None)
+        refresh_storage = self.params.get('refresh_storage', bool)
+        host = find_obj(self.content, [vim.HostSystem], name=esxi_host_name)
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/vmware/vmware_host_scanhba.py]:
```suggestion
        host = find_obj(self.content, [vim.HostSystem], name=esxi_host_name)
        
        if host is None:
           self.module.fail_json(msg="Unable to find host system %s in the given configuration." % esxi_host_name) 
```

File: lib/ansible/modules/cloud/vmware/vmware_host_scanhba.py [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+
+
+def main():
+    argument_spec = vmware_argument_spec()
+    argument_spec.update(
+        esxi_hostname=dict(type='str', required=True),
+        refresh_storage=dict(type='bool', default=False, required=False)
+    )
+    module = AnsibleModule(argument_spec=argument_spec,
+                           supports_check_mode=False)
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/vmware/vmware_host_scanhba.py]: I would recommend to add `check_mode` for this module. You can check other modules in VMware space to see how it works for them.

File: lib/ansible/modules/cloud/vmware/vmware_host_scanhba.py [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+        super(VmwareHbaScan, self).__init__(module)
+
+    def scan(self):
+        esxi_host_name = self.params.get('esxi_hostname', None)
+        refresh_storage = self.params.get('refresh_storage', bool)
+        host = find_obj(self.content, [vim.HostSystem], name=esxi_host_name)
+
+        host.configManager.storageSystem.RescanAllHba()
+        if refresh_storage is True:
+            host.configManager.storageSystem.RefreshStorageSystem()
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/vmware/vmware_host_scanhba.py]: return status from this and handle accordingly in `module.exit_json`

**Review by Akasurde — CHANGES_REQUESTED** [review]
Great addition. I would love to see unit test and integration test for this change. You can check [here](https://github.com/ansible/ansible/tree/devel/test/units/modules/cloud/vmware) for unit and [here](https://github.com/ansible/ansible/tree/devel/test/integration/targets/) for integration test.

**Comment by Akasurde** [comment]
@pdellaert @dericcrago @Im0 @tchernomax @ckotte @jeking3 @Tomorrow9 Could you please review this ?

File: lib/ansible/modules/cloud/vmware/vmware_host_scanhba.py (hunk 7 above)
Comment (mikeeaton83) [reply lib/ansible/modules/cloud/vmware/vmware_host_scanhba.py]: What would you expect it to do in check mode? It either rescans the hbas or it doesn't.. can you elaborate on this?

**Review by pdellaert — APPROVED** [review]
LGTM

Good new addition. 
One (non-required) suggestion is to be able to do it or a whole cluster, not just a single host. An example of how you can do this:
https://github.com/ansible/ansible/blob/6a2aac487dc2ab2f548c67df93cc89ead7f6514a/lib/ansible/modules/cloud/vmware/vmware_host_active_directory.py#L53-L62 

+ 

https://github.com/ansible/ansible/blob/6a2aac487dc2ab2f548c67df93cc89ead7f6514a/lib/ansible/modules/cloud/vmware/vmware_host_active_directory.py#L124-L128

And then loop over the host in the `scan` method.

Just a suggestion, can be done in a second PR after this has merged, of course.

File: lib/ansible/modules/cloud/vmware/vmware_host_scanhba.py (hunk 7 above)
Comment (mikeeaton83) [reply lib/ansible/modules/cloud/vmware/vmware_host_scanhba.py]: I've done as requested.

**Review by pdellaert — APPROVED** [review]
LGTM!

Thanks on the quick addition :)

File: lib/ansible/modules/cloud/vmware/vmware_host_scanhba.py (hunk 2 above)
Comment (Akasurde) [reply lib/ansible/modules/cloud/vmware/vmware_host_scanhba.py:23]: I don't see this change in latest commit.

File: lib/ansible/modules/cloud/vmware/vmware_host_scanhba.py (hunk 5 above)
Comment (Akasurde) [reply lib/ansible/modules/cloud/vmware/vmware_host_scanhba.py:81]: 👍

Comment (mikeeaton83) [reply lib/ansible/modules/cloud/vmware/vmware_host_scanhba.py:81]: This has been done :)

**Comment by if-meaton** [comment]
I've added the integration tests - haven't got time to do the unit tests currently. can this be merged without them?

**Review by michaeldeaton — APPROVED** [review]
Happy to move this PR to my none work GitHub account

**Review by Akasurde — APPROVED** [review]

**Pull request merged**